clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java

```diff
@@ -289,41 +288,191 @@ public void testAppend() throws IOException {
+        ObjectLocation dstObjLoc = fs.pathToObjectLocation(dst);
+        mockExistingFilePath(dstObjLoc);
+
+        boolean renamed = fs.rename(src, dst);
+        Assert.assertTrue(renamed);
+        Assert.assertTrue(dstPathLinkedToSrcPhysicalAddress(srcObjLoc, dstObjLoc));
+        verifyObjDeletion(srcObjLoc);
     }
+
+    /**
+     * file -> existing-directory-name: rename(src.txt, existing-dstdir) -> existing-dstdir/src.txt
+     */
+    @Test
+    public void testRename_existingFileToExistingDirName() throws ApiException, IOException {
+        Path src = new Path("lakefs://repo/main/existing-dir1/existing.src");
+        ObjectLocation srcObjLoc = fs.pathToObjectLocation(src);
+        mockExistingFilePath(srcObjLoc);
+
+        Path fileInDstDir = new Path("lakefs://repo/main/existing-dir2/existing.src");
+        ObjectLocation fileObjLoc = fs.pathToObjectLocation(fileInDstDir);
+        Path dst = new Path("lakefs://repo/main/existing-dir2");
+        ObjectLocation dstObjLoc = fs.pathToObjectLocation(dst);
+        mockExistingDirPath(dstObjLoc, fileObjLoc);
+
+        boolean renamed = fs.rename(src, dst);
+        Assert.assertTrue(renamed);
+        Path expectedDstPath = new Path("lakefs://repo/main/existing-dir2/existing-dir1/existing.src");
+        Assert.assertTrue(dstPathLinkedToSrcPhysicalAddress(srcObjLoc, fs.pathToObjectLocation(expectedDstPath)));
+        verifyObjDeletion(srcObjLoc);
+    }
+
+    /**
+     * rename(srcDir(containing srcDir/a.txt, srcDir/b.txt), non-existing-dstdir) -> non-existing-dstdir/a.txt, non-existing-dstdir/b.txt
+     */
+    @Test
+    public void testRename_existingDirToNonExistingDirName() throws ApiException, IOException {
+        Path fileInSrcDir = new Path("lakefs://repo/main/existing-dir/existing.src");
+        ObjectLocation fileObjLoc = fs.pathToObjectLocation(fileInSrcDir);
+        Path srcDir = new Path("lakefs://repo/main/existing-dir");
+        ObjectLocation srcDirObjLoc = fs.pathToObjectLocation(srcDir);
+        mockExistingDirPath(srcDirObjLoc, fileObjLoc);
+
+        Path dst = new Path("lakefs://repo/main/non-existing-dir");
+        ObjectLocation dstObjLoc = fs.pathToObjectLocation(dst);
+        mockNonExistingPath(dstObjLoc);
+
+        boolean renamed = fs.rename(srcDir, dst);
+        Assert.assertTrue(renamed);
+        Path expectedDstPath = new Path("lakefs://repo/main/non-existing-dir/existing.src");
+        Assert.assertTrue(dstPathLinkedToSrcPhysicalAddress(fileObjLoc, fs.pathToObjectLocation(expectedDstPath)));
+        verifyObjDeletion(fileObjLoc);
+    }
+
+    /**
+     * rename(srcDir(containing srcDir/a.txt), existing-dstdir) -> existing-dstdir/srcDir/a.txt
+     */
+    @Test
+    public void testRename_existingDirToExistingDirName() throws ApiException, IOException {
+        Path fileInSrcDir = new Path("lakefs://repo/main/existing-dir1/existing.src");
+        ObjectLocation srcFileObjLoc = fs.pathToObjectLocation(fileInSrcDir);
```

Comment:
1. Can we have multiple files in the directory here?
2. Should/can we have more than one page's worth of files here (so we can see pagination at work)?